File: src/mdescriptor/core/control.py
"""Cooperative computation control exposed by the core contract."""

from __future__ import annotations

from typing import Any


class _FallbackComputeControl:
    """Fallback shape used only when the native extension is unavailable."""

    def __init__(self) -> None:
        self._cancelled = False
        self._completed = 0
        self._total = 0

    def reset(self, total: int) -> None:
        if isinstance(total, bool) or not isinstance(total, int) or total < 0:
            raise ValueError("control total must be a non-negative integer")
        self._cancelled = False
        self._completed = 0
        self._total = total

    def cancel(self) -> None:
        self._cancelled = True

    def cancelled(self) -> bool:
        return self._cancelled

    def completed(self) -> int:
        return self._completed

    def total(self) -> int:
        return self._total

    def mark_completed(self) -> None:
        self._completed += 1


class ComputeControl:
    """Lazily construct the native control without importing native at package import."""

    def __new__(cls) -> Any:
        try:
            from .. import _native
        except ImportError:
            return _FallbackComputeControl()
        return _native.ComputeControl()


__all__ = ["ComputeControl"]

File: cpp/src/common/control.cpp
#include "mdescriptor/descriptor.hpp"

#include <stdexcept>

namespace mdescriptor {

void ComputeControl::reset(std::int64_t total) {
    if (total < 0) {
        throw std::invalid_argument("control total must be non-negative");
    }
    cancelled_.store(false, std::memory_order_release);
    completed_.store(0, std::memory_order_release);
    total_.store(total, std::memory_order_release);
}

void ComputeControl::cancel() noexcept { cancelled_.store(true, std::memory_order_release); }
bool ComputeControl::cancelled() const noexcept { return cancelled_.load(std::memory_order_acquire); }
std::int64_t ComputeControl::completed() const noexcept { return completed_.load(std::memory_order_acquire); }
std::int64_t ComputeControl::total() const noexcept { return total_.load(std::memory_order_acquire); }
void ComputeControl::mark_completed() noexcept { completed_.fetch_add(1, std::memory_order_acq_rel); }

} // namespace mdescriptor
